### PR TITLE
Add support for multiple groups and users

### PR DESCRIPTION
**Pull request by Kandeel4411**
Currently we only support assigning a single user and group to a check which is quite limiting. This PR aims to improve that by adding support for multiple assignable groups and users

### CODE REVIEW

File: django_datawatch/0005_migrate_new_assigned_fields.py
Comment (lociii) [inline django_datawatch/0005_migrate_new_assigned_fields.py:1]: Guess that file doesn't belong here. Migration is already present in /migrations

File: django_datawatch/admin.py [hunk 1]
```diff
@@ -1,15 +1,23 @@
...
+    extra = 0
+
 @admin.register(Result)
 class CheckAdmin(admin.ModelAdmin):
     list_display = ('slug', 'identifier', 'status')
     readonly_fields = ('created', 'modified')
     search_fields = ('slug', 'identifier', 'payload_description')
-    list_filter = ('status', 'slug', 'assigned_to_group')
-
+    list_filter = ('status', 'slug', 'assigned_groups', 'assigned_users')
```
Comment (lociii) [inline django_datawatch/admin.py:19]: Not sure if it makes sense to add list filters for assigned users, those could be hundreds or thousands.

File: django_datawatch/base.py [hunk 2]
```diff
@@ -131,17 +131,26 @@ def get_form_class(self):
...
+        with transaction.atomic():
+            with track_status_history(self.slug, self.get_identifier(payload), status):
+                # save the check
+                dataset, created = Result.objects.update_or_create(
+                    slug=self.slug, identifier=self.get_identifier(payload),
+                    defaults=defaults)
+
+                # set assigned users and groups
+                if groups := self.get_assigned_groups(payload, status):
+                    assert len(groups) == len(set(g.pk for g in groups)), 'groups must be unique'
```
Comment (lociii) [inline django_datawatch/base.py:148]: I'd also be fine with letting the exception bubble through as customizing a check to yield the same group twice would be considered a misuse of the application logic.

File: django_datawatch/base.py [hunk 3]
```diff
@@ -131,17 +131,26 @@ def get_form_class(self):
...
+                    dataset.assigned_groups.set(groups)
+
+                if users := self.get_assigned_users(payload, status):
+                    assert len(users) == len(set(u.pk for u in users)), 'users must be unique'
```
Comment (lociii) [inline django_datawatch/base.py:152]: Same as for groups.

File: django_datawatch/locale/de/LC_MESSAGES/django.po [hunk 4]
```diff
@@ -72,6 +72,12 @@ msgstr "Ausgeblendet bis"
 msgid "Acknowledge reason"
 msgstr "Ausblendungsgrund"
 
+msgid "Assigned users"
+msgstr "Zugewiesener Benutzer"
```
Comment (lociii) [inline django_datawatch/locale/de/LC_MESSAGES/django.po:76]: `Zugewiesene Benutzer`

File: django_datawatch/locale/de/LC_MESSAGES/django.po [hunk 5]
```diff
@@ -84,6 +90,27 @@ msgstr "Zu Status"
 msgid "Result status history"
 msgstr "Ergebnis-Status-Historie"
 
+msgid "Group"
+msgstr "Gruppen"
```
Comment (lociii) [inline django_datawatch/locale/de/LC_MESSAGES/django.po:94]: `Gruppe`

File: django_datawatch/locale/de/LC_MESSAGES/django.po [hunk 6]
```diff
@@ -84,6 +90,27 @@ msgstr "Zu Status"
...
+msgstr "Zugewiesene Gruppen"
+
+msgid "Group must be unique across the result"
+msgstr "Die Gruppe muss im gesamten Ergebnis eindeutig sein"
+
+msgid "Result assigned user"
+msgstr "Zugewiesener Benutzer"
+
+msgid "Result assigned users"
+msgstr "Zugewiesener Benutzern"
```
Comment (lociii) [inline django_datawatch/locale/de/LC_MESSAGES/django.po:109]: `Zugewiesene Benutzer`

File: django_datawatch/locale/de/LC_MESSAGES/django.po [hunk 7]
```diff
@@ -84,6 +90,27 @@ msgstr "Zu Status"
...
+msgstr "Gruppen"
+
+msgid "Result assigned group"
+msgstr "Zugewiesene Gruppe"
+
+msgid "Result assigned groups"
+msgstr "Zugewiesene Gruppen"
+
+msgid "Group must be unique across the result"
+msgstr "Die Gruppe muss im gesamten Ergebnis eindeutig sein"
```
Comment (lociii) [inline django_datawatch/locale/de/LC_MESSAGES/django.po:103]: `Gruppen müssen je Ergebnis eindeutig sein`

File: django_datawatch/locale/de/LC_MESSAGES/django.po [hunk 8]
```diff
@@ -84,6 +90,27 @@ msgstr "Zu Status"
...
+
+msgid "Result assigned user"
+msgstr "Zugewiesener Benutzer"
+
+msgid "Result assigned users"
+msgstr "Zugewiesener Benutzern"
+
+msgid "User must be unique across the result"
+msgstr "Der Benutzer muss im gesamten Ergebnis eindeutig sein"
```
Comment (lociii) [inline django_datawatch/locale/de/LC_MESSAGES/django.po:112]: `Benutzer müssen je Ergebnis eindeutig sein`

File: django_datawatch/querysets.py [hunk 9]
```diff
@@ -9,8 +9,10 @@
 
 class ResultQuerySet(models.QuerySet):
     def for_user(self, user):
-        return self.filter(Q(assigned_to_group__isnull=True) | Q(assigned_to_group__in=user.groups.all()),
-                           Q(assigned_to_user__isnull=True) | Q(assigned_to_user=user))
+        user_groups = user.groups.all()
+        return self.filter(
+            Q(assigned_groups__in=user_groups) | Q(assigned_users=user),
```
Comment (lociii) [inline django_datawatch/querysets.py:14]: Isn't this now excluding the results that have no user and no group assigned?
Would you mind adding tests for that case?

**Review by lociii — CHANGES_REQUESTED** [review]

File: django_datawatch/0005_migrate_new_assigned_fields.py
Comment (Kandeel4411) [reply django_datawatch/0005_migrate_new_assigned_fields.py:1]: Ah right, I was testing with it and forgot to remove it 👍

File: django_datawatch/tests/test_post_delete.py [hunk 10]
```diff
@@ -39,13 +39,12 @@ def test_update_related_calls_backend(self, mock_get_checks_for_model):
         mock_get_checks_for_model.return_value = [CheckPostDelete]
 
         # mock the manager
-        manager = mock.Mock(spec=ResultQuerySet)
-        Result.objects = manager
```
Comment (Kandeel4411) [inline django_datawatch/tests/test_post_delete.py:43]: I changed this as this was making quite the bug, it was globally affecting all tests as no scope was limited here

File: django_datawatch/querysets.py [hunk 11]
```diff
@@ -9,8 +9,11 @@
 
 class ResultQuerySet(models.QuerySet):
     def for_user(self, user):
-        return self.filter(Q(assigned_to_group__isnull=True) | Q(assigned_to_group__in=user.groups.all()),
-                           Q(assigned_to_user__isnull=True) | Q(assigned_to_user=user))
+        user_groups = user.groups.all()
+        return self.filter(
```
Comment (lociii) [inline django_datawatch/querysets.py:13]: Shouldn't this be
```python
return self.filter(
    Q(assigned_groups__isnull=True)  | Q(assigned_groups__in=user_groups),  # not assigned to a group or to one of mine
    Q(assigned_users__isnull=True) | Q(assigned_users=user),  # not assigned to a user of i'm one of the assignees
)

File: django_datawatch/tests/test_querysets.py [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+User = get_user_model()
+
+class ResultQuerySetTestCase(TestCase):
+    def setUp(self):
+        self.user = User.objects.create_user(**{User.USERNAME_FIELD: 'test_user'})
+        self.group = Group.objects.create(name='test_group')
+
+    def test_for_user(self):
+        result_without_users_and_groups = Result.objects.create(slug='test_slug', identifier='test_identifier')
```
Comment (lociii) [inline django_datawatch/tests/test_querysets.py:15]: We could have all of the following combinations.
Does testing all of these make sense to you?

* No group and user
   * Test positive when user has any group assigned
   * Test positive when user has no groups assigned at all
* One group but no user
   * Test positive when user is part of that group
   * Test negative when user is not part of that group
* Multiple groups but no user
   * Test positive when user is part of one of these groups
   * Test negative when user is not part of any of these group
* One user but no group
   *  Test positive when user is the assigned one
   *  Test positive when user is not the assigned one
* Multiple users but no groups
   * Test positive when user is one of the assigned users
   * Test negative when user isn't one of the assigned users
* One group and one user
   * Test positive when user is part of that group
   * Test positive when user is the assigned one
   * Test negative when user is neither part of that group nor the assigned user
* Multiple groups and one user
   * Test positive when user is part of one of these groups
   * Test positive when user is the assigned one
   * Test negative when user is neither part of these groups nor the assigned user
* One group and multiple users
   * Test positive when user is part of that group
   * Test positive when user is one of the assigned users
   * Test negative when user is neither part of that group nor one of the assigned users

File: django_datawatch/querysets.py (hunk 11 above)
Comment (Kandeel4411) [reply django_datawatch/querysets.py:13]: About this, I am not sure if this is a valid case but I modified what you suggested to count for the case if the user is part of the assigned users but not part of the assigned groups (assigned groups is not null)

Comment (lociii) [reply django_datawatch/querysets.py:13]: Might be that a user was in a group, got a result assigned and then gets removed from that group.  
They should then still be able to see that result.

File: django_datawatch/base.py [hunk 13]
```diff
@@ -131,17 +131,24 @@ def get_form_class(self):
...
+                dataset, created = Result.objects.update_or_create(
+                    slug=self.slug, identifier=self.get_identifier(payload),
+                    defaults=defaults)
+
+                # set assigned users and groups
+                if groups := self.get_assigned_groups(payload, status):
+                    dataset.assigned_groups.set(groups)
+
+                if users := self.get_assigned_users(payload, status):
+                    dataset.assigned_users.set(users)
```
Comment (lociii) [inline django_datawatch/base.py:151]: Please clear if no users returned as this value might be calculated and change over time
```python
dataset.assigned_users.clear()
```

**Review by lociii — CHANGES_REQUESTED** [review]

**Review by lociii — APPROVED** [review]

**Review by lociii — APPROVED** [review]

**Pull request merged**